File: Leetcode_Code/22. Generate Parentheses/22.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        if (n == 0) return {};
        vector<string> result;
        generateParenthesis_dfs(n, "", 0, 0, result);
        return result;
    }

private:
    void generateParenthesis_dfs(int n, string now, int left, int right, vector<string> &result) {
        if (left == n && right == n) {
            result.push_back(now);
            return;
        }
        if (left < n)
            generateParenthesis_dfs(n, now + "(", left + 1, right, result);
        if (left > right)
            generateParenthesis_dfs(n, now + ")", left, right + 1, result);
    }
};

File: Leetcode_Code/22. Generate Parentheses/22.py
class Solution(object):
    def __init__(self):
        self.result = []

    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        if n is not 0:
            self.generateParenthesis_dfs(n, "", 0, 0)
        return self.result

    def generateParenthesis_dfs(self, n, now, left, right):
        if left is n and right is n:
            self.result.append(now)
            return
        if left < n:
            self.generateParenthesis_dfs(n, now + "(", left + 1, right)
        if right < left:
            self.generateParenthesis_dfs(n, now + ")", left, right + 1)
